File: app/src/main/java/info/nukoneko/cuc/android/kidspos/ui/main/calculate/CalculatorLayout.java
package info.nukoneko.cuc.android.kidspos.ui.main.calculate;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import info.nukoneko.cuc.android.kidspos.R;
import info.nukoneko.cuc.android.kidspos.databinding.ViewCalculatorBinding;

public final class CalculatorLayout extends LinearLayout {
    @Nullable
    private Listener mListener = null;

    public CalculatorLayout(Context context) {
        this(context, null, 0);
    }

    public CalculatorLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CalculatorLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        final ViewCalculatorBinding binding =
                DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.view_calculator, this, true);
        binding.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mListener != null) {
                    if (view instanceof CalculatorView) {
                        mListener.onClickNumber(((CalculatorView) view).getNumber());
                    } else if (view.getId() == R.id.calc_num_C) {
                        mListener.onClickClear(view);
                    }
                }
            }
        });
    }

    public void setOnCalculatorClickListener(Listener listener) {
        mListener = listener;
    }

    public interface Listener {
        void onClickNumber(int number);

        void onClickClear(View view);
    }
}

File: app/src/main/kotlin/info/nukoneko/cuc/android/kidspos/ui/main/calculate/CalculatorLayout.kt
package info.nukoneko.cuc.android.kidspos.ui.main.calculate

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.widget.LinearLayout
import info.nukoneko.cuc.android.kidspos.R

class CalculatorLayout : LinearLayout {
    constructor(context: Context) : super(context)
    constructor(context: Context, attr: AttributeSet? = null) : super(context, attr)
    constructor(
        context: Context,
        attr: AttributeSet? = null,
        defStyleAttr: Int = 0
    ) : super(context, attr, defStyleAttr)

    var listener: Listener? = null

    private val onClickListener = OnClickListener { view ->
        when (view) {
            is CalculatorView -> listener?.onNumberClick(view.number)
            else -> {
                if (view.id == R.id.btn_delete_one) listener?.onClearClick()
            }
        }
    }

    private val viewIds: Array<Int> = arrayOf(
        R.id.btn_0, R.id.btn_1, R.id.btn_2,
        R.id.btn_3, R.id.btn_4, R.id.btn_5,
        R.id.btn_6, R.id.btn_7, R.id.btn_8,
        R.id.btn_9, R.id.btn_delete_one
    )

    init {
        val root = LayoutInflater.from(context).inflate(R.layout.view_calculator_layout, this)
        for (viewId in viewIds) {
            root.findViewById<View>(viewId).setOnClickListener(onClickListener)
        }
    }

    interface Listener {
        fun onNumberClick(number: Int)

        fun onClearClick()
    }
}
